File: problem_034.c
// -*- compile-command: "gcc -o problem_034 problem_034.c -Wall" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>

int getFactorial(int input)
{
    switch (input)
    {
    case 0: return 1;
    case 1: return 1;
    case 2: return 2;
    case 3: return 6;
    case 4: return 24;
    case 5: return 120;
    case 6: return 720;
    case 7: return 5040;
    case 8: return 40320;
    case 9: return 362880;
    default: return 0;
    }
}

int main()
{
  printf("Project Euler - Problem 34:\n"
         "Find the sum of all numbers which are equal to the sum of the factorial of their digits.\n\n");

  const int maxNum = 1000000;
  int total = 0;
  int number = 0;

  for (number=3 ; number<maxNum ; number++)
  {
      // Calculate the factorials
      int numberSum = 0;
      int decompose = number;
      while (decompose > 0)
      {
          numberSum += getFactorial(decompose % 10);
          decompose /= 10;
      }

      // Compare
      if ( numberSum == number )
          total += numberSum;
    }

      printf("Sum: %d\n", total);

      return 0;
}

File: problem_034.rs
// -*- compile-command: "rustc -o problem_034_rs problem_034.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 34:");
    println!("Find the sum of all numbers which are equal to the sum of the factorial of their digits.\n");

    // Calculate factorials
    let mut factorial = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    for i in 1..factorial.len() {
        factorial[i] = i * factorial[i-1];
    }

    let mut sum = 0;

    // Find solutions
    for i in 10..100000 {
        let mut n = i;
        let mut fs = 0;
        while n > 0 {
            fs += factorial[n % 10];
            n /= 10;
        }
        if fs == i {
            sum += i;
        }
    }

    println!("{}", sum);
}
